notifications/v1/client.go

```diff
@@ -0,0 +1,158 @@
+package v1
+
+import (
+	"context"
+	"encoding/json"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/lib/background"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"go.uber.org/atomic"
+)
+
+type subscriber struct {
+	logger       log.Logger
+	handlers     SubscriberHandlers
+	subscription *pubsub.Subscription
+
+	// state indicates the state of workers, must be one of "idle", "started" or
+	// "stopped".
+	state *atomic.String
+	// cancelContext is the function to cancel the context of the receiver that
+	// effectively stops the receiver.
+	cancelContext context.CancelFunc
+}
+
+type SubscriberOptions struct {
+	// ProjectID is the GCP project ID that the Pub/Sub subscription belongs to. It
+	// is almost always the same GCP project that the Cloud Run service is deployed
+	// to.
+	ProjectID string
+	// SubscriptionID is the GCP Pub/Sub subscription ID to receive SAMS
+	// notifications from.
+	SubscriptionID string
+	// ReceiveSettings is the settings for receiving messages of the subscription. A
+	// zero value means to use the default settings.
+	ReceiveSettings ReceiveSettings
+	// Handlers is the collection of subscription handlers for each type of SAMS
+	// notifications.
+	Handlers SubscriberHandlers
+}
+
+func (opts SubscriberOptions) Validate() error {
+	if opts.ProjectID == "" {
+		return errors.New("ProjectID is required")
+	}
+	if opts.SubscriptionID == "" {
+		return errors.New("SubscriptionID is required")
+	}
+	return nil
+}
+
+// NewSubscriber creates a new background routine for receiving SAMS
+// notifications from given GCP project ID and Pub/Sub subscription ID.
+//
+// Users should prefer to use the top-level 'sams.NewNotificationsV1Subscriber'
+// constructor instead.
+func NewSubscriber(logger log.Logger, opts SubscriberOptions) (background.Routine, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	client, err := pubsub.NewClient(context.Background(), opts.ProjectID)
+	if err != nil {
+		return nil, errors.Wrap(err, "create GCP Pub/Sub client")
+	}
+	subscription := client.Subscription(opts.SubscriptionID)
+	subscription.ReceiveSettings = opts.ReceiveSettings
+	return &subscriber{
+		logger:       logger.Scoped("notification.subscriber"),
+		handlers:     opts.Handlers,
+		subscription: subscription,
+		state:        atomic.NewString(stateIdle),
+	}, nil
+}
+
+func (s *subscriber) Start() {
+	if !s.state.CompareAndSwap(stateIdle, stateStarted) {
+		s.logger.Error("failed to start subscriber", log.Error(errors.New(`not in "idle" state`)))
```

Comment:
or, maybe we can let this be idempotent, and just return with a no-op

Reply:
+1 to panicing. At this point, there is no way for the system to recover. So just "halt and catch fire" is for the best. (Since otherwise, it's possible for this to maybe make it through unit tests, CI/CD, and explode later than we'd like.)

Reply:
The way it is doing is practically noop (doesn't try to fail anything) but I think also surfacing the error would be enough for the system owner notice something is wrong that they need to fix.

Reply:
Alright, since I got two votes for panic, let's panic :D